### PR TITLE
Fix contracts that contain double type and float operations

### PR DESCRIPTION
**Pull request by aerkol**
- [x] Add TimestampExtensions class to be used in contracts for time operations.
- [x] Replace time operations and other codes in contracts that cause double type usage and float operations in IL codes.
- [x] Fix remaining audit errors in contracts
- [x] Enable contract auditor (replace AssemblyChecker with the new one)
- [x] Use TimestampHelper when getting current time.

### CODE REVIEW

File: src/AElf.Contracts.Election/ElectionContract_Election.cs [hunk 1]
```diff
@@ -209,14 +209,14 @@ public override Empty Vote(VoteMinerInput input)
...
-            var lockDays = (input.EndTimestamp - Context.CurrentBlockTime.ToTimestamp()).ToTimeSpan().TotalDays;
+            var lockDays = (input.EndTimestamp - Context.CurrentBlockTime.ToTimestamp()).Days();
             Assert(lockDays >= State.MinimumLockTime.Value,
                 $"Invalid lock time. At least {State.MinimumLockTime.Value} {(TimeUnit) State.BaseTimeUnit.Value}");
             Assert(lockDays <= State.MaximumLockTime.Value,
                 $"Invalid lock time. At most {State.MaximumLockTime.Value} {(TimeUnit) State.BaseTimeUnit.Value}");
 
             State.LockTimeMap[Context.TransactionId] =
-                GetTimeSpan(input.EndTimestamp.ToDateTime(), Context.CurrentBlockTime);
+                GetTimeSpan(input.EndTimestamp, Context.CurrentBlockTime.ToTimestamp());
```
Comment (loning) [inline src/AElf.Contracts.Election/ElectionContract_Election.cs]: directly change the context interface

Comment (aerkol) [reply src/AElf.Contracts.Election/ElectionContract_Election.cs]: To confirm, I will change Context.CurrentBlockTime type from DateTime to Timestamp?

**Comment by aerkol** [comment]
Building contracts with overflow check helps passing the contract auditor validations, however it somehow breaks some of the test cases that shows the signs that there is an overflow during the tests. Disabling overflow check now and not enabling build with overflow check in contracts. This is to be fixed later.

**Comment by aerkol** [comment]
**Summary**
- We need to implement IL code processing monitoring to prevent DOS completely.
- We need to fix failing test cases when contracts are built with "check for arithmetic overflow", then we can enable float ops validator in contract auditor. Currently all contracts passes this check once they are built with -checked flag but unfortunately test cases fail. This points out to an existing overflow in contracts with those test cases, further investigation is required.
- Preventing float op codes brings some minor disadvantages such as contract devs can't use foreach to iterate through arrays (ienumerables are fine).

File: src/AElf.Contracts.AssociationAuth/AElf.Contracts.AssociationAuth.csproj [hunk 2]
```diff
@@ -20,4 +20,8 @@
     <ItemGroup>
       <ProjectReference Include="..\AElf.Sdk.CSharp\AElf.Sdk.CSharp.csproj" />
     </ItemGroup>
+    <Target Name="CopyContractDll" AfterTargets="AfterBuild">
```
[the remark this answers is not in the document]
Comment (aerkol) [reply src/AElf.Contracts.AssociationAuth/AElf.Contracts.AssociationAuth.csproj]: Added the copy task to current targets file

File: bench/AElf.Benchmark/Fibonacci16Tests.cs [hunk 3]
```diff
@@ -55,7 +55,7 @@ public async Task Fibonacci16()
                     BlockHeight = _chain.BestChainHeight
                 },
                 _transaction,
-                DateTime.UtcNow);
+                DateTime.UtcNow.ToTimestamp());
```
Comment (loning) [inline bench/AElf.Benchmark/Fibonacci16Tests.cs]: change to Helper method, do not directly use DateTime.UtcNow

File: contract/AElf.Contracts.Consensus.AEDPoS/Types/MinerList.cs [hunk 4]
```diff
@@ -32,7 +33,7 @@ public partial class MinerList
                 minerInRound.PublicKey = sortedMiners[i];
                 minerInRound.Order = i + 1;
                 minerInRound.ExpectedMiningTime =
```
Comment (loning) [inline contract/AElf.Contracts.Consensus.AEDPoS/Types/MinerList.cs]: do not need change TimeStamp to DateTime, directly use TimeStamp + duration

File: src/AElf.Sdk.CSharp/SafeDateTime.cs [hunk 5]
```diff
@@ -0,0 +1,146 @@
+using System;
+using Google.Protobuf.WellKnownTypes;
+
+namespace AElf.Sdk.CSharp
+{
+    public class SafeDateTime
```
Comment (loning) [inline src/AElf.Sdk.CSharp/SafeDateTime.cs]: remove safe date time, only let smart contract use Timestamp and Duration, because DateTime has DateTime.Kind and double

File: contract/AElf.Contracts.Consensus.AEDPoS/Types/MinerList.cs (hunk 4 above)
Comment (aerkol) [reply contract/AElf.Contracts.Consensus.AEDPoS/Types/MinerList.cs]: Should I add extensions for Timestamp? Like AddMilliseconds, AddSeconds then use duration to add inside the extensions?

**Comment by loning** [comment]
we can merge and change it later

**Comment by aerkol** [comment]
> we can merge and change it later

Sure, adding DateTimeHelper now, will be pushing soon

File: bench/AElf.Benchmark/BlockExecutingTests.cs [hunk 6]
```diff
@@ -51,7 +52,7 @@ public async Task IterationSetup()
                     ChainId = chain.Id,
                     Height = chain.BestChainHeight + 1,
                     PreviousBlockHash = chain.BestChainHash,
-                    Time = Timestamp.FromDateTime(DateTime.UtcNow)
+                    Time = DateTimeHelper.Now.ToTimestamp()
```
Comment (loning) [inline bench/AElf.Benchmark/BlockExecutingTests.cs]: TimestampHelper.Now, I remembered some one has an implement

Comment (aerkol) [reply bench/AElf.Benchmark/BlockExecutingTests.cs]: Sure, I will use that one

**Comment by aerkol** [comment]
Fixed conflicts now fixing the remaining build errors
